clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -1611,4 +1611,208 @@ mlir::Value ScalarExprEmitter::VisitAbstractConditionalOperator(
 
         patchVoidOrThrowSites();
       });
-}
\ No newline at end of file
+}
+
+mlir::Value ScalarExprEmitter::VisitBinLAnd(const clang::BinaryOperator *E) {
+  if (E->getType()->isVectorType()) {
+    llvm_unreachable("NYI");
+  }
+
+  bool InstrumentRegions = CGF.CGM.getCodeGenOpts().hasProfileClangInstr();
+  mlir::Type ResTy = ConvertType(E->getType());
+  mlir::Location Loc = CGF.getLoc(E->getExprLoc());
+
+  // If we have 0 && RHS, see if we can elide RHS, if so, just return 0.
+  // If we have 1 && X, just emit X without inserting the control flow.
+  bool LHSCondVal;
+  if (CGF.ConstantFoldsToSimpleInteger(E->getLHS(), LHSCondVal)) {
+    if (LHSCondVal) { // If we have 1 && X, just emit X.
+
+      mlir::Value RHSCond = CGF.evaluateExprAsBool(E->getRHS());
+
+      if (InstrumentRegions) {
+        llvm_unreachable("NYI");
+      }
+      // ZExt result to int or bool.
+      return Builder.createZExtOrBitCast(RHSCond.getLoc(), RHSCond, ResTy);
+    }
+    // 0 && RHS: If it is safe, just elide the RHS, and return 0/false.
+    if (!CGF.ContainsLabel(E->getRHS()))
+      return Builder.getBool(false, Loc);
+  }
+
+  mlir::Value LHSCondV = CGF.evaluateExprAsBool(E->getLHS());
+  mlir::Value RHSCondV = CGF.evaluateExprAsBool(E->getRHS());
```

Comment:
The `RHSCondV` should be evaluated inside the first ternary expression, otherwise there's no short-circuiting to be done here.

Try building this testcase:
```
typedef enum {
  A = 0,
  B = 1,
  C = 2,
  D = 3,
  Last = 0xF
} enumy;

bool t(enumy e) {
    return e == A || e == B || e == C || e == D;
}
```

You gonna see that `cir.cmp`s that need to execute conditionally are being early evaluated. Please also add a test for this in the testcases.